### PR TITLE
Redis CR LCM

### PR DESCRIPTION
**Pull request by Ghostbaby**
Cover the lifecycle management for the Redis CR, including:

- [x] Provision
- [x] Delete
- [x] Scale
- [x] Update
- [x] Status
- [ ] Upgrade
- [ ] Backup
- [ ] Restore

Fixes https://github.com/goharbor/harbor-cluster-operator/issues/4

### CODE REVIEW

File: controllers/cache/generate.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+		Version:  "v1",
+		Resource: "redisfailovers",
+	}
+)
+
+func (d *defaultCache) generateRedisCR() (*unstructured.Unstructured, error) {
+	name := d.Request.Name
+	nameSpace := d.Request.Namespace
+
+	resource := utils.GetRequests("1", "1Gi")
```
Comment (Ghostbaby) [inline controllers/cache/generate.go:26]: After the harbor cluster operator spec is updated, this place needs to be modified.

File: controllers/cache/reconcile.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	ExpectCR *unstructured.Unstructured
+	ActualCR *unstructured.Unstructured
+	Labels   map[string]string
+}
+
+type defaultCache struct {
+	ReconcileCache
+}
+
+func (d *defaultCache) Reconcile() (*harborCluster.CRStatus, *reconciler.Results) {
```
Comment (Ghostbaby) [inline controllers/cache/reconcile.go:49]: This is Redis main reconcile.

File: controllers/cache/client/api/v1/types.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
+package v1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// RedisFailover represents a Redis failover
+type RedisFailover struct {
```
[the remark this answers is not in the document]
Comment (Ghostbaby) [reply controllers/cache/client/api/v1/types.go:9]: ok, I will directly import RedisFailover CRD.

File: api/v1/harborcluster_types.go [hunk 4]
```diff
@@ -60,3 +60,19 @@ type HarborClusterList struct {
 func init() {
 	SchemeBuilder.Register(&HarborCluster{}, &HarborClusterList{})
 }
+
+type CRStatus struct {
```
[the remark this answers is not in the document]
Comment (steven-zou) [reply api/v1/harborcluster_types.go:64]: Is `ExternalService` enough? We may need to return other information.

File: config/certmanager/certificate.yaml [hunk 5]
```diff
@@ -17,8 +17,8 @@ metadata:
 spec:
   # $(SERVICE_NAME) and $(SERVICE_NAMESPACE) will be substituted by kustomize
   dnsNames:
-  - $(SERVICE_NAME).$(SERVICE_NAMESPACE).svc
-  - $(SERVICE_NAME).$(SERVICE_NAMESPACE).svc.cluster.local
+    - $(SERVICE_NAME).$(SERVICE_NAMESPACE).svc
```
Comment (steven-zou) [inline config/certmanager/certificate.yaml:20]: Which format tool are you using?

File: controllers/cache/deploy.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"log"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	harborCluster "src/github.com/goharbor/harbor-cluster-operator/api/v1"
+)
+
+func (d *defaultCache) Deploy(status *harborCluster.CRStatus) (*harborCluster.CRStatus, error) {
```
Comment (steven-zou) [inline controllers/cache/deploy.go:15]: The status argument is designed for what purpose?

File: controllers/cache/deploy.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+// DeploySecret deploy the Redis Password Secret
+func (d *defaultCache) DeploySecret() error {
+	secret := &corev1.Secret{}
+	name := d.Request.Name
+	namespace := d.Request.Namespace
+	sc := d.generateRedisSecret(d.Labels)
+	if err := controllerutil.SetControllerReference(d.Harbor, sc, d.Scheme); err != nil {
+		return err
+	}
+	err := d.Client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: namespace}, secret)
```
Comment (steven-zou) [inline controllers/cache/deploy.go:64]: Use `context.TODO()` or sharing the context of harbor-cluster reconcile controller?

File: controllers/cache/generate.go (hunk 1 above)
Comment (steven-zou) [reply controllers/cache/generate.go:26]: It should be created based on the Redis section of cluster spec.

File: controllers/cache/reconcile.go (hunk 2 above)
Comment (steven-zou) [reply controllers/cache/reconcile.go:49]: Please add comments when necessary.

**Comment by steven-zou** [comment]
# Overall thoughts

## Let's discuss

I'm thinking of the interface of LCM controllers of the dependent CRs. We have `HarborCluster` Spec, maybe we can explicitly define an lcm controller interface to cover the logic of different stages of the dependent service components.

```go
pkg lcm


type Scaler interface {
    // Scale up
    Up(newReplicas uint64)(*CRStatus, error)

    // Scale down
    Down(newReplicas uint64)(*CRStatus, error)
}

type Controller interface {
    // Provision the new dependent service by the related section of cluster spec.
    Provision(spec *HarborCluster)(*CRStatus, error)

    // Delete the service
    Delete()(*CRStatus, error)

    // Scale the service
    Scale()Scaler

    // Update the service
    Update(spec *HarborCluster) (*CRStatus, error)

    // More...
}
```

Each dependent service can also implement the `Reconciler` interface base on the `lcm. Controller` interface.

```go

type Reconciler interface {
    // Reconcile the dependent service.
    Reconcile()(*CRStatus, error)
}
```

Create a reference object to keep the references of all dependent services.

```go
type ServiceGetter interface {
    // For Redis
    Cache(spec *HarborCluster)cache.Reconciler

    // For database
    Database(spec *HarborCluster)db.Reconciler

    // For storage
    Storage(spec *HarborCluster)storage.Reconciler

    // For harbor itself
    Harbor(spec *HarborCluster)harbor.Reconciler
}
```

For cluster reconcile controller,

```go
  type Controller struct{
      // Add top ref?
      Services ServiceGetter
  }


  func (c *Controller) Reconcile() (Enqueue, error) {
      var spec *HarborCluster
      // ...
      cs, err:= c.Services.Cache(spec).Reconcile()
      if err!= nil {
          // Return
      }

      // Update status
      err = c.UpdateStatus(cs)
  }
```

@Ghostbaby @wangcanfengxs @soulseen

File: controllers/cache/deploy.go (hunk 7 above)
Comment (Ghostbaby) [reply controllers/cache/deploy.go:64]: ok, It has been modified, use the context of harbor-cluster reconcile controller.

File: config/certmanager/certificate.yaml (hunk 5 above)
Comment (Ghostbaby) [reply config/certmanager/certificate.yaml:20]: I use default goland file watchers config, already change to use `go fmt`.

File: Dockerfile [hunk 8]
```diff
@@ -1,4 +1,4 @@
-# Build the manager binary
+w# Build the manager binary
```
[the remark this answers is not in the document]
Comment (Ghostbaby) [reply Dockerfile:1]: already fixed.

**Comment by steven-zou** [comment]
```go
type Spec struct {
  old *HarborCluster
  new *HarborCluster
}
```

**Comment by steven-zou** [comment]
```go
type SpecComparator interface {
   
}
```

File: api/v1/harborcluster_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (steven-zou) [reply api/v1/harborcluster_types.go:64]:
```go
type CRStatus struct {
  //...
  Properties []*Property
 //...
}
```

```go

const (
  ProperPass = "Password"
  ProperConn = "Connection"
  // add more ...
)
```

```go
type Property struct {
   Name string // const values
  // Parse.get(Type).Parse(Value) int/string
   Type string //  int, string, secret , getter utility methods will be provided 
   Value interface{} // string
}
```

**Pull request closed without merging**